field/generator/generator_test.go

```diff
@@ -73,32 +78,58 @@ func TestIntegration(t *testing.T) {
 	moduli["e_nocarry_edge_0127"] = "170141183460469231731687303715884105727"
 	moduli["e_nocarry_edge_1279"] = "10407932194664399081925240327364085538615262247266704805319112350403608059673360298012239441732324184842421613954281007791383566248323464908139906605677320762924129509389220345773183349661583550472959420547689811211693677147548478866962501384438260291732348885311160828538416585028255604666224831890918801847068222203140521026698435488732958028878050869736186900714720710555703168729087"
 
+	assert.NoError(GenerateCommonASM(2, asmDir, false))
+	assert.NoError(GenerateCommonASM(3, asmDir, false))
+	assert.NoError(GenerateCommonASM(7, asmDir, false))
+	assert.NoError(GenerateCommonASM(8, asmDir, false))
+
 	for elementName, modulus := range moduli {
 		var fIntegration *field.FieldConfig
 		// generate field
 		childDir := filepath.Join(rootDir, elementName)
 		fIntegration, err = field.NewFieldConfig("integration", elementName, modulus, false)
-		if err != nil {
-			t.Fatal(elementName, err)
-		}
-		if err = GenerateFF(fIntegration, childDir); err != nil {
-			t.Fatal(elementName, err)
-		}
+		assert.NoError(err)
+		assert.NoError(GenerateFF(fIntegration, childDir, asmDir, "../../../asm"))
 	}
 
 	// run go test
 	wd, err := os.Getwd()
 	if err != nil {
 		t.Fatal(err)
 	}
-	packageDir := filepath.Join(wd, rootDir) + string(filepath.Separator) + "..."
-	cmd := exec.Command("go", "test", packageDir)
-	if err := cmd.Run(); err != nil {
-		if exitErr, ok := err.(*exec.ExitError); ok {
-			t.Fatal(string(exitErr.Stderr))
-		} else {
-			t.Fatal(err)
+	packageDir := filepath.Join(wd, rootDir) // + string(filepath.Separator) + "..."
+
+	// list all subdirectories in package dir
+	var subDirs []string
+	err = filepath.Walk(packageDir, func(path string, info os.FileInfo, err error) error {
+		if info.IsDir() && path != packageDir {
+			subDirs = append(subDirs, path)
 		}
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	errGroup := errgroup.Group{}
+
+	for _, subDir := range subDirs {
+		// run go test in parallel
+		errGroup.Go(func() error {
+			cmd := exec.Command("go", "test")
+			cmd.Dir = subDir
+			var stdouterr strings.Builder
+			cmd.Stdout = &stdouterr
+			cmd.Stderr = &stdouterr
+			if err := cmd.Run(); err != nil {
```

Comment:
This approach doesn't kill other `go test` commands. Instead, we could create the new errGroup with `errgroup.WithContext` and provide the context to `cmd := exec.CommandContext`. The command should be killed when the errgroup context is cancelled (which is done when any `errGroup.Go` calls returns a non-nil error.

Or we could leave it as is - the test returns anywait later and it means the child processes should be killed afaik.